kernel/src/table_changes/scan_file.rs

```diff
@@ -0,0 +1,403 @@
+//! This module handles [`CDFScanFile`]s for [`TableChangeScan`]. A [`CDFScanFile`] consists of all the
+//! metadata required to generate a change data feed. [`CDFScanFile`] can be constructed using
+//! [`CDFScanFileVisitor`]. The visitor reads from engine data with the schema [`cdf_scan_row_schema`].
+//! You can convert engine data to this schema using the [`get_cdf_scan_row_expression`].
+use itertools::Itertools;
+use std::collections::HashMap;
+use std::sync::{Arc, LazyLock};
+
+use super::log_replay::TableChangesScanData;
+use crate::actions::visitors::visit_deletion_vector_at;
+use crate::engine_data::{GetData, TypedGetData};
+use crate::expressions::{column_expr, Expression};
+use crate::scan::state::DvInfo;
+use crate::schema::{
+    ColumnName, ColumnNamesAndTypes, DataType, MapType, SchemaRef, StructField, StructType,
+};
+use crate::utils::require;
+use crate::{DeltaResult, EngineData, Error, RowVisitor};
+
+#[allow(unused)]
+pub(crate) struct UnresolvedCDFScanFile {
+    pub scan_file: CDFScanFile,
+    pub remove_dvs: Arc<HashMap<String, DvInfo>>,
+}
+#[allow(unused)]
+#[derive(Debug)]
+pub(crate) struct ResolvedCDFScanFile {
+    pub scan_file: CDFScanFile,
+    pub selection_vector: Option<Vec<bool>>,
+}
+
+// The type of action associated with a [`CDFScanFile`].
+#[allow(unused)]
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) enum CDFScanFileType {
+    Add,
+    Remove,
+    Cdc,
+}
+
+/// Represents all the metadata needed to read a Change Data Feed.
+#[allow(unused)]
+#[derive(Debug, PartialEq, Clone)]
+pub(crate) struct CDFScanFile {
+    /// The type of action this file belongs to. This may be one of add, remove, or cdc.
+    pub scan_type: CDFScanFileType,
+    /// a `&str` which is the path to the file
+    pub path: String,
+    /// a [`DvInfo`] struct, which allows getting the selection vector for this file
+    pub dv_info: DvInfo,
+    /// a `HashMap<String, String>` which are partition values
+    pub partition_values: HashMap<String, String>,
+    /// the commit version that this action was performed in
+    pub commit_version: i64,
+    /// the timestamp of the commit that this action was performed in
+    pub commit_timestamp: i64,
+}
+
+pub(crate) type CDFScanCallback<T> = fn(context: &mut T, scan_file: CDFScanFile);
+
+/// Transforms an iterator of TableChangesScanData into an iterator  of
+/// `UnresolvedCDFScanFile` by visiting the engine data.
+#[allow(unused)]
+pub(crate) fn scan_data_to_scan_file(
+    scan_data: impl Iterator<Item = DeltaResult<TableChangesScanData>>,
+) -> impl Iterator<Item = DeltaResult<UnresolvedCDFScanFile>> {
+    scan_data
+        .map(|scan_data| -> DeltaResult<_> {
+            let scan_data = scan_data?;
+            let callback: CDFScanCallback<Vec<CDFScanFile>> =
+                |context, scan_file| context.push(scan_file);
+            let result = visit_cdf_scan_files(
```

Comment:
I don't understand why this method can't take `scan_data` directly (instead of breaking it up and passing (almost) all its fields as individual args? Then we could directly probe the remove_dvs while visiting adds, and `CDFScanFile` could hold dv_info for both add and remove, and the next PR would have a much easier time resolving DVs (no need to pass the hashmap to it).

Is it attempting to preserve similarity to the analagous `visit_scan_files` or something?

Reply:
Good point, I was preserving the similarity with `visit_scan-files`. iirc this part of the code has ffi-related reasons for being the way that it is 🤔  I'll look into it, but I'll ping @nicklan if he has input

Reply:
I ended up switching it and it simplified it nicely. Ig we can deal with engine support later given that we aren't targeting that rn